Basic ipv6 support

sebageek opened the pull request:
There is still some testing to be done. Also, the tests are not yet complete

sebageek committed:
Call update only once in VBInterface postflight

In VBInterface.postflight() we only need to call _update(), as we're
already scoped to a router (and have a context).

sebageek committed:
IPv6 support for internal networks

With this patch we're supporting IPv6 partly in our infrastructure, at
least for internal networks. This enables the v6 address family on our
VRFs (if there is an IPv6 subnet connected to the router) and we
configure IPv6 addresses on interfaces where needed.

For BGPVPN we currently skip all v6 subnets. For extraroutes there is
still testing to be done.

sebageek committed:
IPv6 support for router routes

We can now configure IPv6 routes on our hardware routers, including
default gateways.

As we now return an empty dict if there are no routes we should no
longer see any route diffs when the router has no default route
configured.

sebageek committed:
IPv6 support for BGPVPN code

Support announcing internal IPv6 CIDRs via BGPVPN, if we have them. This
does not yet support route-targets / route-maps inside VRFs.

sebageek committed:
IPv6 support for prefix lists

For our external prefixes we need support for IPv6 prefix lists.

sebageek committed:
IPv6 support for vrf address-family v6

We now support setting route-targets and export maps on the v6
address-family.

sebageek committed:
IPv6 support for Route Maps

Route Maps can now reference IPv6 prefix lists. We also did some
maintenance on the RouteMap class, like removing the unused enable_bgp
flag and making sure that specifying both prefix_list and access_list
don't override each other's match conditions (though this hasn't been
used so far).

As we don't know yet how we'll transport IPv6 prefixes from ACI/EVPN to
our routers we'll for now leave the RoutePrefix to be v4 only and don't
reference it for v6 addresses. Once we know how that part of the infra
is going to work we'll fill it with life!

sebageek committed:
Properly remove prefix seqs in preflight()

With the introduction of the new prefix list yang model in c626065 we
also needed to introduce a cleanup method in preflight() that cleans up
stale entries of a prefix list. This removal did not work properly, as
we provided a seq as a string (number), while the Prefix class expects a
PrefixSeq yang model. Therefore the cleanup in preflight failed with:

AttributeError: asr1k_neutron_l3.models.netconf_yang.prefix in
to_delete_dict 'str' object has no attribute 'no'

When we convert the prefix list to a proper model, cleanup works.

sebageek committed:
IPv6 support for external interfaces

We can now configure external interfaces with IPv6, including DAPNet
support. This includes a rework of how the Router() class handles IP
addresses and offers route redistribution via route-maps + prefix lists
instead of explicit network statements.

The InterfaceList now has three new methods to get the list of IPs for a
router that are external, internal or routable. Routable IPs are defined
as the list of internal IPs where the address scope of the subnet
matches the address scope of the gateway (external) interface.

The BGP AddressFamily class now supports advertising of routes via
route-map. This means we have a route-map, which references prefix lists
that contain routable and internal IPs + their extra routes. For
routable IPs and their extraroutes we set a route target: routable IPs
get the cloud vrf's id, routable extraroutes have a prepended 1 to the
vrf id. This is very much the same as it is currently handled on the
device, though the config was pre-provisioned and only referenced via
route-maps on the respective network statements. This feature is
currently only enabled for IPv6 default and can be enabled for IPv4 with
the config option advertise_bgp_ipv4_routes_via_redistribute. The reason
of doing this is that editing the BGP config tree currently locks up the
whole device config due to requiring a full config resync.

Prefix lists referenced by route-maps that do not contain any prefix
don't appear as config on the device. The referencing route-map treats
non-existing prefix-lists as "always matching". As our expectation of
how this would work was empty prefix-list means "no match", we now have
to adjust our expectations of this feature and always add an entry if
the list would be empty otherwise. This is done via add_deny_if_empty,
which adds a deny for everything. The seq 4242 was deliberately chosen
to not collide with any of our existing rules, not because we saw any
problems with this, but because we didn't want to end up in a situation
where the device locks up because "there is already a rule with that
seq", even though we generally think this should work.

mutax commented on asr1k_neutron_l3/models/netconf_yang/vrf.py:
```diff
@@ -233,16 +242,18 @@ def postflight(self, context, method):
+            routes_v4 = VrfRouteV4.get_for_vrf(context=context, vrf=self.id)
+            routes_v6 = VrfRouteV6.get_for_vrf(context=context, vrf=self.id)
+            routes = routes_v4 + routes_v6
+            if routes:
```
`if not routes` or `if routes`?

sebageek replied:
The former.

mutax commented on asr1k_neutron_l3/models/netconf_yang/vrf.py:
```diff
@@ -263,17 +274,18 @@ def postflight(self, context, method):
                 result = af._delete(context=context)
                 LOG.debug(result)
-                LOG.info("Deleted hanging address family in vrf {} postflight.".format(self.name))
+                LOG.debug("Deleted hanging address family in vrf %s postflight.", self.name)
```
add `af` here as well?

mutax commented on asr1k_neutron_l3/common/utils.py:
```diff
@@ -134,6 +134,14 @@ def ip_in_network(ip, net):
     return IPAddress(ip) in IPNetwork(net)
 
 
+def get_ip_version(ip):
```
would rename to `get_ip_addr_version` or `get_ip_net_version`, or something along those lines,  get_ip_version sounds somehow generic to me.  (nitpick)

sebageek replied:
Currently this method accepts both an ip network or address and regardless of what it is it'll return the version. I think I'd keep it, as both proposed names would suggest it wouldn't take the other format.

sven-rosenzweig commented on asr1k_neutron_l3/models/neutron/l3/interface.py:
```diff
@@ -52,59 +52,114 @@ def all_interfaces(self):
+    def get_external_cidrs_v6(self):
+        return self.get_external_cidrs(6)
+
+    def get_routable_networks(self, ip_version):
```
`_get_routable_networks` to mark those method private?

sebageek replied:
Used externally by Router class

sven-rosenzweig commented on asr1k_neutron_l3/models/neutron/l3/interface.py:
```diff
@@ -52,59 +52,114 @@ def all_interfaces(self):
+
+        # routable networks are networks for which the subnet's subnet pool address scope
+        # matches the address scope of the external gateway
+        scope_key = {4: "address_scope_v4", 6: "address_scope_v6"}[ip_version]
```
Do we need to catch an `KeyError` here?

sebageek replied:
We hardcode the 4 and 6 everywhere (or loop over a hardcoded list) so I think this is fine. If it breaks, we need to fix it, so handling this would be "look into sentry, fix it in the code". I don't  think we need to handle this case here by catching that exception.

mutax replied:
..until IPv8 comes along 😸

sven-rosenzweig commented on asr1k_neutron_l3/models/neutron/l3/interface.py:
```diff
@@ -52,59 +52,114 @@ def all_interfaces(self):
+        if not self.gateway_interface:
+            return []
+
+        return [subnet['cidr']
```
From my point of view this is easier to understand: 
```
 subnets = []
        for interface in self.internal_interfaces:
            for subnet in interface.subnets:
                if subnet.get('cidr') and utils.get_ip_version(subnet['cidr']) == ip_version:
                    subnets.append(subnet['cidr'])
```

sebageek replied:
Did you mean to reference L56-59 with the double-for in `get_internal_cidrs()`? If so, I can also roll this out, if this is easier to understand. I find double-fors in a list-comprehension come quite easy, once you roll them out in your head, but I agree that they might not be the most common thing to see. I can roll it out in this case if it helps the understandability and we'll discuss it on a case-by-case basis.

sven-rosenzweig commented on asr1k_neutron_l3/models/neutron/l3/interface.py:
```diff
@@ -52,59 +52,114 @@ def all_interfaces(self):
 
         return result + self.internal_interfaces + self.orphaned_interfaces
 
+    def get_internal_cidrs(self, ip_version):
```
` def _get_internal_cidrs(self, ip_version):` mark this as private?

sebageek replied:
It's used in several places in the models.neutron.l3.router.Router class, where we have the ip version in a variable anyways. Why should I make it private?

sebageek committed:
No dynamic NAT for routers without external gw

Routers that do not have an external gateway also don't have a bridge
domain associated with it, meaning the dynamic NAT list would always
produce a diff. Therefore we delete the dynamic NAT list if there's
nothing to configure and also make sure it doesn't appear in the diff.

mutax commented on asr1k_neutron_l3/models/netconf_yang/bgp.py:
```diff
@@ -44,79 +46,49 @@ class BGPConstants(object):
-            else:
-                dict[BGPConstants.VRF] = OrderedDict()
-                dict[BGPConstants.VRF]["id"] = asn
+        dict = cls._remove_base_wrapper(dict, context)
```
Would it make sense to use this opportunity to rename the parameter from `dict` to something that does not conflict with an internal datatype? Each time I see this code again after some time,  I stop and need some cycles to process what's going on there. Also my IDE complains :)

sebageek replied:
This opportunity... probably not. But this is also something that has annoyed me for a long time. If we do this, I'd like to do it for all of the `remove_wrapper()` methods at the same time - just in case something is using that as a keyword argument somewhere.

mutax commented on asr1k_neutron_l3/models/netconf_yang/bgp.py:
```diff
@@ -136,48 +108,181 @@ def _wrapper_preamble(self, dict, context):
         device_af = self._internal_get(context=context)
         if device_af and device_af.vrf:
-            return super(AddressFamily, self)._delete_no_retry(context, *args, **kwargs)
+            return super()._delete_no_retry(context, *args, **kwargs)
```
nitpick: what do we think regarding an explicit `return None` in such cases?

sebageek replied:
"If we start returning something, we should return something in every path". As I only changed the super(), I didn't touch anything else here. I'll add one in this case :)

mutax commented on asr1k_neutron_l3/models/netconf_yang/bgp.py:
```diff
@@ -196,9 +301,33 @@ def cidr(self):
+        return cls(number=cidr, route_map=route_map)
+
+    @property
+    def cidr(self):
```
would it make sense to add `from_cidr()` and `cidr()` to `class NetworkV4` as well, so they provide the same functionality? Maybe adding a common parent class NetworkBase, like with some of the others?
(nitpick)

sebageek replied:
I see a `from_cidr()` and `cidr()` for in my `NetworkV4` and I think they're indeed used in a way that they provide the same interface to the NetworkV4/V6 class. Introducing a common baseclass could be done to have a defined interface, but the methods have subtle differences, so we couldn't share the code.

mutax commented on asr1k_neutron_l3/models/netconf_yang/route.py:
```diff
@@ -87,23 +90,21 @@ def get_primary_filter(cls, **kwargs):
     @classmethod
     def remove_wrapper(cls, dict, context):
-        dict = super(VrfRoute, cls)._remove_base_wrapper(dict, context)
+        dict = cls._remove_base_wrapper(dict, context)
```
same question again - can we rename the variable from dict to something else?

sebageek replied:
Let's do it in a followup

mutax commented on asr1k_neutron_l3/models/netconf_yang/route.py:
```diff
@@ -87,23 +90,21 @@ def get_primary_filter(cls, **kwargs):
-        super(VrfRoute, self).__init__(**kwargs)
+        return {
+            self.IP_KEY: {
+                self.LIST_KEY: dict,
```
dict variable shadowing builtin (added note here for completeness)

mutax commented on asr1k_neutron_l3/models/neutron/l3/prefix.py:
```diff
@@ -13,70 +13,79 @@
+        if add_deny_if_empty and not prefixes:
+            self._rest_definition.add_seq(
+                # use 4242 as a likely not-used seq no
+                self._make_seq(4242, self.DEFAULT, action="deny")
```
make 4242 a centrally defined constant?

sebageek replied:
Hmm I don't see a benefit in it yet. We don't use it anywhere else. I think I'd keep it as-is for now.

mutax commented on asr1k_neutron_l3/models/neutron/l3/route.py:
```diff
@@ -19,41 +19,61 @@
 
 
-class RouteCollection(base.Base):
+class BaseRouteCollection(base.Base):
```
I think with most (all?) of the newly introduced Base-classes for v4/v6, "Base" was added at the end of the existing class name, so to follow that "naming scheme" this class would be called `RouterCollectionBase` .  That also fits nicely with `RouteCollectionV{4,6}`

mutax commented on asr1k_neutron_l3/models/neutron/l3/route.py:
```diff
@@ -19,41 +19,61 @@
+
+
+class RouteV6(base.Base):
+    def __init__(self, router_id, destination, nexthop):
```
Would it be possible (and make sense) to have RouteV4 and RouteV6 accept the same parameters in their constructor? Then they could also get a common parent class and avoid a bit of code duplication.

sebageek replied:
Incorporated that, as discussed in Slack

mutax commented on asr1k_neutron_l3/models/neutron/l3/route_map.py:
```diff
@@ -31,28 +31,29 @@ def __init__(self, name, rt=None, routable_interface=False):
-        sequences.append(route_map.MapSequence(seq_no=seq, operation='deny', prefix_list='ext-{}'.format(self.vrf)))
+        sequences.append(route_map.MapSequence(seq_no=seq, operation='deny', prefix_list=f'ext-{self.vrf}'))
+
+        # ipv6
```
to follow the scheme I would suggest adding a seq+=10 in line 53 and move the one from line 55 to after the append, but within the if branch. (possible nitpick)

mutax commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -222,40 +225,53 @@ def _build_nat_acl(self):
+            self.interfaces.get_external_cidrs_v4() +
+            self.interfaces.get_external_cidrs_v6() +
+            self.interfaces.get_internal_cidrs_v4() +
+            self.interfaces.get_internal_cidrs_v4()
```
copy paste error, the last one should be v6?

sebageek replied:
Yes, most definitely! Good find!

mutax commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -297,17 +313,77 @@ def _build_prefix_lists(self):
+        extraroutes_rt = None
+        if self.rt and ':' in self.rt:
+            # 65126:106 --> 65126:1106
+            asn, sf = self.rt.split(":", 2)
```
Should the 2 be a 1?

sebageek replied:
100%, we always want to specify the amount of splits we wanna do, not the number of results we want :)

mutax commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -297,17 +313,77 @@ def _build_prefix_lists(self):
+
+            # redistribute routemaps
+            route_map.Redist4RouteMap(self.router_id, self.rt, extraroutes_rt,
+                                      enabled=(self.enable_ipv4 and
```
I have seen `self.enable_ipv4 and cfg.CONF.asr1k_l3.advertise_bgp_ipv4_routes_via_redistribute` several times now, might improve readability and reduce change for mistakes to add a class variable or property for that? (nitpick)

sebageek replied:
This is the only place where this is used. The other two places have `ip_version` in a local variable and don't depend on `self.enable_ipv4` (in some cases the class itself has a `delete()` hook if the AF is disabled and therefore still needs to be created). So sadly I can't easily summarize them.

mutax commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -459,21 +527,24 @@ def _delete(self):
+        results.append(self.routes[4].delete())
+        results.append(self.routes[6].delete())
 
         for key in self.dynamic_nat.keys():
```
could also iterate over `values()` and drop the `get()`

sebageek replied:
Yes, but I'd move this to a follow up / code cleanup run. In this PR I only tried to clean up what was hindering me in developing the IPv6 feature.

sven-rosenzweig commented on asr1k_neutron_l3/models/netconf_yang/vrf.py:
```diff
@@ -161,8 +165,12 @@ def to_dict(self, context):
 
-        if self.address_family_ipv4 is not None:
-            definition[VrfConstants.ADDRESS_FAMILY][VrfConstants.IPV4] = self.address_family_ipv4.to_dict(context)
+        if self.address_family_ipv4 or self.address_family_ipv6:
```
```
 af = definition[VrfConstants.ADDRESS_FAMILY] = {}
 if self.address_family_ipv4 is not None:
                af[VrfConstants.IPV4] = self.address_family_ipv4.to_dict(context)
 if self.address_family_ipv6 is not None:
                af[VrfConstants.IPV6] = ""
```

Why do we need the outer if-clause?

sebageek replied:
We need to make sure not to override the whole address family dict, but we also only want to set the ADDRESS_FAMILY key if we have an address family to set. So first, we check if we have any, then create the entry and then fill that up with whatever address families are available. We had cases where we just had the inner ifs and if both were true we'd override the first entry with the second entry. In this case, we'd have a `definitionVrfConstants.ADDRESS_FAMILY] = ...` in both cases.

sven-rosenzweig commented on asr1k_neutron_l3/models/neutron/l3/interface.py:
```diff
@@ -65,6 +66,7 @@ def __init__(self, router_id, router_port, extra_atts):
         self._primary_subnet_id = None
         self.ip_address = self._ip_address()
```
For consistency reasons, does it make sense to rename this to  `self.ip4_addresses`?

sebageek replied:
I think that has been done in a later commit.

sven-rosenzweig commented on asr1k_neutron_l3/models/netconf_yang/route.py:
```diff
@@ -68,7 +68,10 @@ class VrfRoute(NyBase):
     VRF_XPATH_FILTER = "/native/ip/route/vrf[name='{vrf}']"
```
We could set this variable to ``None`` as this is overrriten in ```VrfRouteV6 or VrfRouteV4`

sven-rosenzweig commented on asr1k_neutron_l3/models/netconf_yang/route.py:
```diff
@@ -113,37 +114,41 @@ def neutron_router_id(self):
 
-        if isinstance(self.routes, list):
+        vrf_routes = []
+        if self.routes:
```
Do we really need to check twice as we have``if not self.routes:``?

sebageek replied:
....no, that doesn't make sense at all.

sven-rosenzweig commented on asr1k_neutron_l3/models/neutron/l3/prefix.py:
```diff
@@ -32,48 +34,72 @@ def __init__(self, name_prefix, router_id, gateway_interface, internal_interface
         if self.gateway_interface is not None:
             i = 1
             for subnet in sorted(self.gateway_interface.subnets, key=itemgetter('id')):
+                if utils.get_ip_version(subnet.get('cidr')) != self.IP_VERSION:
```
Does it make sense to check whether `self.IP_VERSION `is not ``None``?

sebageek replied:
As this is a class variable we have under our control I'd expect that to always be either 4 or 6.

mutax commented on asr1k_neutron_l3/models/neutron/l3/nat.py:
```diff
@@ -53,23 +53,20 @@ def _rest_definition(self):
+            self.bridge_domain = self.gateway_interface.bridge_domain
 
     @property
     def _rest_definition(self):
```
I know you did not touch the bottom of that method, but I think adding a `return None` could improve readability so one does not overlook that there is an `elseif` and None can be a valid return value. (that should link to the property _rest_definition)

sebageek replied:
I think I'd leave it like it is for now. We have two cases there and we don't have any explicit handling of `None` values. We could think about using enums or making sure specific values are not allowed here, but before doing that we're probably doing the `dict` refactoring. ;)

sebageek committed:
IPv6 support for ext iface loop prevention

We want to replicate the loop prevention implemented in
a63bf9432a680a1bebd4a8cd888dee77b02f1c73 for IPv4. This code should
prevent traffic coming from the external interface from leaving the
external interface again, if the vrf decides it needs to leave via the
default route. This can happen when the device the router is connected
to (e.g. ACI) thinks a destination is in a vrf, but the vrf doesn't have
the destination anymore (due to removed fips, removed subnets or
whatever). With IPv4 we had an explicit "set precedence routine" for
internal packets, but when trying to configure this on the device cli it
looks like this is omitted as being the default config. As we don't need
to replicate default config we'll leave that part as-is.

This change results in the following cli change for routers with an ipv6
address on the external interface:

interface BD-VIF 9999
+ ipv6 traffic-filter common ACL-EXT-TOS-V6 out
+ ipv6 policy route-map RM-EXT-TOS-V6

The access-list ACL-EXT-TOS-V6 and route-map RM-EXT-TOS-V6 need to be
pre-provisioned.

sebageek committed:
Add first sync data tests for db to yang model

We now have the first few tests that can create a router in the
OpenStack database and generate sync data from this, which can then be
used to instantiate the models used in the agent that then generate the
netconf yang config for our devices. This means we can more or less now
do end-to-end tests for DB-to-YANG tests, including coverage for if the
Router() class correctly handles the data gotten from the DB and makes
the right decisions.

The current mock-implementation is a best effort approach to get us
started with having tests. As the sync data is normally transferred via
RPC / rabbitmq and then augmented with the address scope info we need to
run some extra steps in the tests in get_sync_data() to have everything
similar to what we have in production.

mutax commented on asr1k_neutron_l3/models/netconf_yang/l3_interface.py:
```diff
@@ -434,3 +465,38 @@ def to_dict(self, context):
+    @classmethod
+    def __parameters__(cls):
+        return [
+            {"key": 'direction'},
```
Is it intentionally to use string literals here and not the constants defined? I wouldn't be surprised it there is a reason - if yes what is the criteria?

sebageek replied:
There is no good reason for this.

The constants have always been there and are used in every `to_dict()` method, but they are nearly never used in the `__parameters__()` method. This is how I got the codebase and how I largely carried on the work. When writing new classes I in some cases used the constants, but while doing so I also noticed that in some cases this became very tedious: It works well for the `key` paramter, but it made the `yang-path` parameter completely unreadable. So, while the constants help against typos, it made it much harder to validate the `yang-path` against new firmware versions. Therefore I largely kept the way it was originally written and instead decided to introduce some tests for the classes.

I think it could make sense to replace the key argument with the already existing constants in some cases - wherever it doesn't hinder the readability, but we don't yet have anything policywise, so currently it's more or less a case-by-case decision.

mutax commented on asr1k_neutron_l3/tests/common/fixtures.py:
```diff
@@ -0,0 +1,231 @@
+                               return_value={'fake-agent': 0}):
+            with self.router(admin_state_up=admin_state_up, tenant_id=tenant_id,
+                             external_gateway_info=external_gateway_info, **kwargs) as router:
+                pass
```
adding a comment what this is doing / why this single pass is required would be highly appreciated by this reviewer but is not required. something like check if a router can be created without raising an exception

sebageek replied:
This is a common pattern used in nearly all Neutron test code (common to Neutron, not common in general). The `with` is not really doing anything on entry/exit and I have no good idea why they use it this way expect that it might "look good". The original code can be found [here](https://github.com/sapcc/neutron/blob/266ae8ad642f37b72404655e7a70ded00fc3e2dc/neutron/tests/unit/extensions/test_l3.py#L469-L472).

As it is such a common pattern I'd leave it uncommented here.

sebageek committed:
Improve VRF AddressFamily handling

We want to be able to add and remove the address-family of a VRF as
needed, so now we only enable it, if we also have an ip address for it -
at least in the case of IPv6. For IPv4 we - for now - always keep the AF
enabled to make sure we don't have any problems with the config
transition. Once we are at a point were we make all device config
dependant on the enable_ipv4 flag we can also think about removing the
AF from teh VRF.

We have a special case in the netconf-yang pulled from the device: an
empty address-family config results in an empty tag, e.g. <ipv6/>, which
then results in this config not appearing in the diff. A diff is needed
for the object to actually be updated, so we add an extra flag to the
Vrf netconf class, that creates an empty address family object if we
find an empty tag on the device.

mutax commented on asr1k_neutron_l3/models/netconf_yang/l3_interface.py:
```diff
@@ -434,3 +465,38 @@ def to_dict(self, context):
+        ]
+
+    def to_dict(self, context):
+        return {L3Constants.PREFIX: self.prefix.lower()}
```
using `self.ITEM_KEY` instead feels cleaner, also in case at some point somebody creates a subclass.

sebageek replied:
So far I've been trying to go with the constants in most cases and only use the `ITEM_KEY` if its needed by a subclass and is actually changing its value. I doubt we will ever subclass this one and I find the `PREFIX` easier to read... But to be honest I could go either way here, both options feel acceptable to me. Do you feel a benefit for the `ITEM_KEY` solution?

mutax replied:
The only benefit I would see is making clear that the key in the dictionary is the same as the definition in the class - but I have no strong opinion. did a quick check, other classes don't use the `ITEM_KEY` as well. lets keep it that way, then its consistent in the file

mutax commented on asr1k_neutron_l3/models/netconf_yang/vrf.py:
```diff
@@ -263,17 +283,18 @@ def postflight(self, context, method):
+            afs = afs_v4 + afs_v6
 
             if len(afs) == 0:
                 LOG.info("No address fammilies to clean")
```
could fix the typo in families

mutax commented on asr1k_neutron_l3/models/netconf_yang/vrf.py:
```diff
@@ -327,6 +348,14 @@ def to_dict(self, context):
         return dict(address_family)
 
 
+class IpV4AddressFamily(IpAddressFamilyBase):
```
hmm. We now have `IpV4AddressFamily` and `AddressFamilyV4` in this file, same for v6.
Is this too confusing on the long run or are we fine with that?
What is our opinion on importing AddressFamilyV4 as AddressFamilyV4_bgp or give the classes defined in this file a more distinctive name? `VrfIPv4AddressFamily` looks ugly, though

sebageek replied:
Let's import just the module and not import the bgp classes into our module namespace. Makes it more readable :)

mutax commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -115,19 +114,13 @@ def _to_global_vrf_id(self, rt):
-                if iface.primary_subnet and 'cidr' in iface.primary_subnet]
+    @property
+    def enable_ipv6(self):
+        return any(iface.ipv6_addresses for iface in self.interfaces.all_interfaces)
```
will this trigger on fe80:: or ::1 on localhost? (not sure what is part of all_interfaces - does it query the router or neutron?)

sebageek replied:
This information comes from Neutron, meaning whatever the user has specified. So it says "has the user requested anything that looks like v6?" --> we enable it.

mutax commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -222,40 +223,53 @@ def _build_nat_acl(self):
+                has_routable_interface = bool(routable_networks)
+                redistribute_map = None
+
+            enable_af = self.enable_ipv4 if ip_version == 4 else self.enable_ipv6
```
(nitpick/more of a comment)
I know the chance that we will ever see something else than v4 or v6 in our lifetime are slim, but somehow not checking for ip_version==6 triggers my inner monk, also because it looks assymetrical to me as if something was missing 
`enable_af = self.enable_ipv4 and ip_version == 4 or self.enable_ipv6 and ip_version == 6`
would be an alternative and at least my brain processes it faster without thinking twice.

sebageek replied:
We have many places in which we make this exact distinction and I think I'd like to keep it this way. I understand your Inner Monk, but then I'd rather have us introduce an enum type for the AF and make sure we always have a 4 or 6 in there. Also, we don't have any handling for if this is not 4 or 6 in the way you propose, so my preference would be to leave it as it is for now.

mutax commented on asr1k_neutron_l3/models/neutron/l3/router.py:
```diff
@@ -297,17 +311,77 @@ def _build_prefix_lists(self):
+                # to avoid potential config locks, we won't configure them
+                continue
+
+            af_enabled = self.enable_ipv4 if ip_version == 4 else self.enable_ipv6
```
same as above

mutax commented on asr1k_neutron_l3/models/netconf_yang/vrf.py:
```diff
@@ -233,16 +251,18 @@ def postflight(self, context, method):
-            LOG.error("Failed to delete {} routes in VRF {} postlight : {}".format(len(routes), self.id, e))
+                LOG.debug("Deleted hanging route %s in vrf %s postflight.", route.name, self.name)
+        except Exception as e:
+            LOG.error("Failed to delete %s routes in VRF %s postlight: %s", len(routes), self.id, e)
```
typo light -> flight

also, as we are catching all exceptions here, do we want to log it as LOG.exception, so we get the stacktrace in the log as well?

sebageek replied:
Corrected the typo. I presume we don't need the stack trace here, as it will most likely fail due to a device error, from which we only need the message to know what's up. Error reporting in the driver could be refactored, mid-term, though.

sebageek committed:
Make IPv6 and NAT pool work together

When providing multiple IP addresses on router create or router update,
we used to take all of them in consideration for NAT pooling. This
prevents user from updating their router with an IPv6 address (as "a
router with two addresses is not allowed for NAT pooling") and also will
fail when trying to create a NAT pool that contains IPv6 addresses. Now
we split all provided addresses in two lists, one per address family.
All IPv4 addresses go the regular NAT pooling / no NAT pooling path. For
IPv6 we allow zero or one address. Providing more than one IPv6 address
will show a nice error message to the user.

toanju commented on asr1k_neutron_l3/models/neutron/l3/vrf.py:
```diff
@@ -34,18 +34,30 @@ def __init__(self, name, description=None, asn=None, rd=None, routable_interface
 
-        self.enable_bgp = False
-        self.map = "exp-{}".format(self.name)
+        # FIXME: hmmm might also need to be AF aware, but... when do we have a routable interface?
```
```suggestion
        # NOTE: hmmm might also need to be AF aware, but... when do we have a routable interface?
```

sebageek replied:
Uhm yeah, the comment in the code is just plain wrong. `self.map` is only used for v4 and v6 has its own map. I think that comment was added before this was the case (when I was still trying to figure out if I can do this with one route-map, which we could, but decided against it, due to other participants regarding this as a more complex solution).

I'll remove the offending comment completely.

toanju commented on asr1k_neutron_l3/models/netconf_yang/vrf.py:
```diff
@@ -263,17 +283,18 @@ def postflight(self, context, method):
                 result = af._delete(context=context)
                 LOG.debug(result)
-                LOG.info("Deleted hanging address family in vrf {} postflight.".format(self.name))
+                LOG.debug("Deleted hanging address family %s in vrf %s postflight.", af, self.name)
```
could be merged with the previous LOG.debug

toanju commented on asr1k_neutron_l3/models/neutron/l3/bgp.py:
```diff
@@ -21,17 +21,22 @@
+        self.redistribute_map = redistribute_map
         self.routable_networks = routable_networks
-        self.networks_v4 = set()
+        self.networks = set()
```
any reason why this is not a list from the start?

sebageek replied:
Deduplication / distrust of the input parameters we get. Just to make sure there are no duplicate networks in there, which would result in yang trying to reconfigure the same object again and again, as "my lists looks different than what's on the device".

toanju commented on asr1k_neutron_l3/models/neutron/l3/interface.py:
```diff
@@ -51,49 +52,115 @@ def all_interfaces(self):
+            for subnet in interface.subnets:
+                if subnet.get('cidr') and utils.get_ip_version(subnet['cidr']) == ip_version:
+                    cidrs.append(subnet['cidr'])
+        cidrs.sort()
```
out of curiosity, do we need to have this sorted for any API or is this for the readability of a human being?

sebageek replied:
I sort these lists, as order is not guaranteed by OpenStack and is in fact considered to be unordered, but when this is the source for a prefix list or interface configuration, order might suddenly matter. Most classes do a sort themselves, but just on the off-chance that we missed it somewhere I like to keep it sorted.

In the past we had some prefix lists that changed their order on each update, causing a lot of config changes for "nothing has changed".

toanju replied:
whee thanks for the details

sven-rosenzweig commented on asr1k_neutron_l3/plugins/l3/service_plugins/l3_extension_adapter.py:
```diff
@@ -405,8 +405,29 @@ def _update_router_gw_info(self, context, router_id, info,
+
+        for ext_ip in ext_ips:
+            ip_version = None
+            if 'subnet_id' in ext_ip:
```
Why do we need to fetch the ip version from the subnet? Can't we rely on the ip_address itself?

From my understanding, the info dict looks like: 
` "external_fixed_ips": [{"subnet_id": "uuid", "ip_address": "<ip-address>"}]`

sven-rosenzweig replied:
Why not 
```
if 'ip_address' in ext_ip:
    blabla
 elif 'subnet_id' in ext_ip:
   blabla
```

sebageek replied:
The `info` dict we're receiving here is not from the OpenStack database, but from the API request of the user, e.g. if the user types in `os router set $router_id --external-gateway $extnet --fixed-ip ip-address=$ip --fixed-ip subnet=$subnet` - this will result in an external gateway interface with two ips, one directly specified, one allocated from the subnet. we will only be having one of those entries in there in nearly all of the cases (I'm not sure adding both would be possible, but it would still be redundant). I can flip the if statement though, to prefer the IP if we happen to have both though.

sven-rosenzweig replied:
thanks for the detailed explanation. Not need to flip this from my side

mutax reviewed (approved):
lgtm
